### PR TITLE
Handle all the new distinct TOC suffixes

### PR DESCRIPTION
**Pull request by p3lim**
This patch also correctly sets the path for these TOC files

See:
https://github.com/Stanzilla/WoWUIBugs/issues/68#issuecomment-830351390
https://github.com/Stanzilla/WoWUIBugs/issues/68#issuecomment-889431675

Properly solves what #103 attempted to do.

(I swear I will not accidentally merge this one)

### CODE REVIEW

**Comment by nebularg** [comment]
> We could consider a breaking change whereby if the first TOC from the result of the find/sort contains a client-specific suffix then the packager exits with an error indicating a potential ambiguity, and requests that the author specify an explicit package name in the .pkgmeta.
> _Originally posted by @Meorawr in https://github.com/BigWigsMods/packager/issues/104#issuecomment-894652168_

I think I like this more than potentially eating part of the addon name until the dev noticed.

Also, I don't see how this properly fixes #103, as the game version set has no influence on which toc file is used. e.g, If you have `MyAddon-Classic.toc` and `MyAddon-BCC.toc`, running `release.sh -g classic` will still the first match, which is `MyAddon-BCC.toc`

**Comment by p3lim** [comment]
> I think I like this more than potentially eating part of the addon name until the dev noticed.

Me too, although it's out of the scope of this PR.

> Also, I don't see how this properly fixes #103, ...

Fair enough, I could account for this, although it'd require a larger refactor.

**Comment by p3lim** [comment]
The whole `## Interface-<game_type>` support added to the packager could probably be deprecated, since it has been resolved by Blizzard themselves.

**Comment by nebularg** [comment]
> Me too, although it's out of the scope of this PR.

I brought it up because Meo closed their PR in response to this 

> Fair enough, I could account for this, although it'd require a larger refactor.

I have a branch which handles multiple tocs much better.  It's my v2, but CurseForge is taking forever to support multiple versions so it's languished a bit. I'll do some testing with it to make sure it does identical builds then I'll push it to let @Stanzilla and @Meorawr break it. 

In the mean time would just unconditionally using a lone toc file work for your workflow?

As an aside, does CurseForge still requires the fallback toc file?

> The whole `## Interface-<game_type>` support added to the packager could probably be deprecated, since it has been resolved by Blizzard themselves.

Once CurseForge supports it, sure 😜My plan is to add an option to create multiple tocs by processing the fallback toc multiple times (done in v2), then to either remove it entirely or, just make that the default behavior down the line (v3)

**Comment by Meorawr** [comment]
> The whole `## Interface-<game_type>` support added to the packager could probably be deprecated, since it has been resolved by Blizzard themselves.

This one's tricky.

While we do now have the per-client file suffixes, we'd probably find that for the common case the only difference between TOC files for each client is going to be the Interface directive; the file list and everything else will _probably_ be identical, with an exception for a few larger addons that have library dependencies that only apply to specific clients.

The `Interface-<type>` directive is useful in that sense since it allows authors to maintain a single TOC file and not have to worry about synchronizing changes for any other directives to - potentially - two other TOC files. This is especially important for things like SavedVariables or the file listing, since you wouldn't want to accidentally forget changes to those.

What I'd like to see is a processing stage which would, given a standard TOC file, generate client-specific TOC files for each client that has an explicit `Interface-<type>` directive - so given the following `MyAddOn.toc`:

```
## Interface-Mainline: 90100
## Interface-BCC: 20502
## Interface-Classic: 11400
## SavedVariables: SomeVariable

File1.lua
File2.lua
```

The tool in question would create `MyAddOn-{Mainline,BCC,Classic}.toc` with the correct "Interface" directive for each listed client, and remove the `MyAddOn.toc` file. If the tool wanted to be particularly fancy, it could allow any directive to be suffixed with `-{Mainline,BCC,Classic}` - though I'm not sure on the usefulness of that. As an example, here's how the generated `MyAddOn-Mainline.toc` would look from the above example:

```
## Interface: 90100
## SavedVariables: SomeVariable

File1.lua
File2.lua
```

My view is that while it'd be convenient for the packager to do this for us, it's also a bit out-of-scope of "packaging". The tool itself could, however, be trivially implemented as a separate community action. I think making stages of the packaging process more modular was a direction @nebularg was interested in at one point, and this would be a good first candidate.

With such a tool in place I'd agree the packager then deprecate its support of these directives. My expectation is that, once deprecated, the `-g <type>` flag would cause the packager to search for and process either a client-specific TOC file of the named type(s), or the generic one, and wouldn't bother attempting to search for or process anything else.

---

As an aside I think the topic of what we want from the packager in terms of multi-client support is probably a bit underspecified at present. I think the tool described above at least takes most of the work out of the packagers' domain and greatly simplifies things, but one thing I would like to see the packager improve upon is the support for multiple game versions in the `-g` flag (#98).

What I'd like is to be able to say `-g mainline,bcc,classic` and - assuming all upload hosts are involved - for the packager to do the following for me automatically:

- **For hosts that support tagging a single release as supporting multiple clients**, generate and upload a single zip to them. The supported game versions should be inferred from the `Interface` directive present within client-specific TOC files.
- **For other hosts that need client-specific zips**, generate one zip per client with TOCs for other clients removed. The game version would be parsed from the `Interface` directive in the remaining TOC file. This should get past the issue of some hosts (CF) not allowing duplicate uploads.

In each case above the packager should only look for and process the Interface directive alone from client-specific TOC files; the author, if they elect to use `-g` with multiple named game flavours, would be expected to provide a matching set of client-specific TOC files that matches the list of clients passed to `-g`. If any are missing then the packager should error.

The fallback TOC wouldn't be processed at all, and thus no need to support per-client Interface directives in this process; authors that want that behaviour can just opt-in to the previously discussed tool and run it prior to the packager to generate the necessary TOCs for the packager to consume.

With the above we'd effectively solve [this common usage](https://github.com/WeakAuras/WeakAuras2/blob/99f584c9710f4dc81ed9f26c4a41c3ec60558686/.github/workflows/build.yml#L37-L58) of the packager by being able to just run it once with `-g <flavours>` and have it do the right thing. One thing that isn't attempted to be solved by this is how client-specific blocks are processed, but I think those are more or less a dying breed in favour of runtime checks for the client flavour at this point anyway.

**Comment by nebularg** [comment]
I pushed my [multitoc support](https://github.com/BigWigsMods/packager/tree/multitoc) branch. With a few [notes](https://github.com/BigWigsMods/packager/tree/multitoc#whats-new-with-v2) on changes in the README. Would be great to get some eyeballs on it. Curse seems to support multiple versions now (with the caveat that a fallback toc file is required), so I should probably get this polished up.

Reading though @Meorawr comments again, I feel like using `-g` should really be a debug tool at this point. The packager will set game versions on the package based on the toc files it finds. I'm not sure in what cases you'd want to change that behavior.  Adding the ability to use `-g bcc,classic` in one step and `-g retail` in the next to restrict versions to those game types wouldn't be terribly hard to add I think, but it seems like an awkward use case.

**Comment by Meorawr** [comment]
Looks good at a quick glance, I'll play with it today if I get some time and see if I can break it some.

Shame that CF still needs a fallback TOC file though.

**Comment by Meorawr** [comment]
As the packager in that branch prefers to generate `_TBC.toc` and `_Vanilla.toc` with the `-S` flag set, it might make sense to have the `## Interface-<Client>`  directives that it pulls from also support that naming scheme, such as the following:

```toc
## Interface-Vanilla: 11401
## Interface-TBC: 20502
## Interface-Retail: 90105

FileA.lua
```

Currently if you try to package the above (via `release.sh -S`) it'll fail to correctly generate TOCs for Vanilla/TBC, resulting in the following:

```
Copying files into .release/MyAddon_Vanilla:
  Copying: MyAddon_Vanilla.toc
    Creating MyAddon_Vanilla_Mainline.toc [90105]
    Creating MyAddon_Vanilla [11401]
    Creating MyAddon_Vanilla_TBC.toc [20502]
```

```sh
$ tree .release/MyAddon_Vanilla
.release/MyAddon_Vanilla
├── CHANGELOG.md
├── MyAddon_Vanilla
├── MyAddon_Vanilla.toc
└── MyAddon_Vanilla_Mainline.toc
```

**Comment by nebularg** [comment]
> As the packager in that branch prefers to generate `_TBC.toc` and `_Vanilla.toc` with the `-S` flag set

I initially used `-(Mainline|Classic|BCC).toc`, but switched to what I thought was the "official" suffixes. Is this what we want? Personally, I prefer the look of hypens but /shrug.

> it might make sense to have the `## Interface-<Client>` directives that it pulls from also support that naming scheme

That was actually suppose to work, but I missed translating the type back to my internal ones 😓

**Comment by Meorawr** [comment]
I think standardising on the "official" suffixes is best long-term, yes.

**Comment by Meorawr** [comment]
Regarding client-specific preprocessor blocks (`@version-retail@`) and such, should the packager perhaps error or at least raise a loud diagnostic if these are detected when doing a build for a project that has multiple client TOCs?

At the minute if I've got `FileA.lua` with the following:

```lua
--@version-retail@
print(1)
--@end-version-retail@
--@version-bcc@
print(1)
--@end-version-bcc@
--@version-classic@
print(1)
--@end-version-classic@
```

Packaging via `release.sh -S` with one multi-interface-lined TOC results in the following being generated in the build:

```lua
--@version-retail@
print(1)
--@end-version-retail@
--[====[@version-bcc@
print(1)
--@end-version-bcc@]====]
--[====[@version-classic@
print(1)
--@end-version-classic@]====]
```

**Comment by nebularg** [comment]
Not sure how i feel about a hard error, but it probably makes sense considering it's for the packager to parse (and will otherwise probably be missed)

Is there a wowpedia page with info on the constants i could link to or something? feels like dumping them in there would be a bit too wordy. I came up with something like this

```
Copying files into .release/MyAddon:
  Copying: FileA.lua
    Error! Build type version keywords found and are not allowed for a multi-version build.
           These should be replaced with lua conditional statements:
           1:--@version-retail@
           4:--@version-bcc@
           7:--@version-classic@
```

**Comment by Ketho** [comment]
> Is there a wowpedia page with info on the constants i could link to or something? feels like dumping them in there would be a bit too wordy.

There is https://wowpedia.fandom.com/wiki/WOW_PROJECT_ID

**Comment by StormFX** [comment]
I imagine it's probably best to use the "official" ToC names, but honestly, I prefer the hyphen over the underscore, too. More often than not, add-ons that are modules of other add-ons or plug-ins for those add-ons are going to be using an underscore as part of their name. Hyphens would make it a bit more distinct.

**Comment by StormFX** [comment]
Just to mention what was discussed on the UI Discord, it would be ideal if, when splitting, to have the main ToC file be assigned to the version specified in the default `Interface:` line. Eg:

```toc
## Interface: 90105
## Interface-TBC: 20502
## Interface-Vanilla: 11401
```

Would yield only three files, `AddOn.toc`, `AddOn_TBC.toc` and `AddOn_Vanilla.toc`.

Also:

- Does the built-type keyword logic apply to ToC files? I'm asking because when splitting, it would be nice to be able to exclude certain lines from generated ToC files in which the line isn't applicable. Eg, I load LibDualSpec directly in the ToC for Retail, but not for the others. If it's too much of a hassle or something, I can just use manual ToC files, but figured I'd mention it. 😋 

- For splitting, can we get a switch or something that tells the packager to use hyphens instead of underscores? 😄

**Comment by nebularg** [comment]
> Edit: Is this usable or just in a review stage? Just wondering. :)

Yea, everything should work fine, I'm just trying to find edge cases and polish it a bit.  The branch will eventually get merged into master and I'll tag it as v2

> Does the built-type keyword logic apply to ToC files?

Yea, people are already building game specific TOCs from running the packager multiple times so it seemed like using that to make all the TOC files at once would be useful.

> For splitting, can we get a switch or something that tells the packager to use hyphens instead of underscores? 😄

You can create your own files using hyphens and whatever suffixes, the packager will read everything supported, but I think I'll leave it using Blizzard's preferred way.

**Comment by StormFX** [comment]
> Yea, everything should work fine, I'm just trying to find edge cases and polish it a bit. The branch will eventually get merged into master and I'll tag it as v2

Nice! I'll have to keep an eye out for the merge so that I don't miss the switch from `@multitoc` to `@master`.

> Yea, people are already building game specific TOCs from running the packager multiple times so it seemed like using that to make all the TOC files at once would be useful.

Right, I was just wondering if would be possible to process the keywords in ToCs for those cases where some lines may need to be in certain versions but not others (eg, libraries that are only compatible with certain versions, etc). That's fine, though. It'd probably be more sane on my end to maintain manual ToCs anyhow. Thanks, though!

> You can create your own files using hyphens and whatever suffixes, the packager will read everything supported, but I think I'll leave it using Blizzard's preferred way.

That's fine. Damn them and their underscores! Thanks for all your time and effort, Neb, we really appreciate it! ❤️

**Comment by Gethe** [comment]
What is the plan for `-g`?

At the moment it doesn't seem to be working correctly in some circumstances. If I have an addon with separated TOCs and run `release.sh -g classic` it seems to ignore the AddOn_Vanilla.toc and package based solely on AddOn.toc. This would then end in an error because that fallback only has version directives for retail, which don't match to classic.

**Comment by nebularg** [comment]
Why are you passing g at all?  The goal is not needing to do anything special and it all just works, so what is happening that you're trying to fix? The purpose of `-g` is back to being a override if toc and version numbers are out of sync. 

In your example, it should being tagging the package with whatever version your classic toc uses, which seemed to work fine in my testing, but if it's using the retail version that's strange. Addon link?

**Comment by Gethe** [comment]
I'm doing this with Aurora https://github.com/Gethe/Aurora, which has different files between retail, vanilla, and tbc. The last action log was using `master`, but I haven't got it to work on `multitoc` either in local tests.
https://github.com/Gethe/Aurora/actions/runs/1493234262

After some additional testing, I found that the issue I'm getting on `multitoc` stems from some old local changes renaming Aurora-Classic.toc to Aurora_Vanilla.toc and -BCC to _TBC. With these changes, neither file gets included in the package regardless of what options I use.
